src/core/Sample.cpp

```diff
@@ -0,0 +1,584 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include "Sample.h"
+
+#include "AudioEngine.h"
+#include "Engine.h"
+#include "SampleBuffer.h"
+
+#include <iostream>
+#include <cassert>
+
+namespace lmms
+{
+    // values for buffer margins, used for various libsamplerate interpolation modes
+    // the array positions correspond to the converter_type parameter values in libsamplerate
+    // if there appears problems with playback on some interpolation mode, then the value for that mode
+    // may need to be higher - conversely, to optimize, some may work with lower values
+    static auto s_interpolationMargins = std::array<f_cnt_t, 5>{64, 64, 64, 4, 4};
+
+    Sample::Sample(std::shared_ptr<SampleBuffer> buffer) :
+        m_buffer(buffer),
+        m_playMarkers({0, static_cast<f_cnt_t>(m_buffer->size()), 0, static_cast<f_cnt_t>(m_buffer->size())}),
+        m_markerSampleRate(m_buffer->sampleRate()) {}
+
+    Sample::Sample(const QString& audioFile) :
+        Sample(std::make_shared<SampleBuffer>(audioFile)) {}
+
+    auto Sample::play(
+        sampleFrame* dst, PlaybackState* state,
+        fpp_t frames, float freq,
+        LoopMode loopMode
+    ) -> bool
+    {
+        const auto [startFrame, endFrame, loopStartFrame, loopEndFrame] = m_playMarkers;
+
+        if (endFrame == 0 || frames == 0)
+        {
+            return false;
+        }
+
+        // variable for determining if we should currently be playing backwards in a ping-pong loop
+        bool isBackwards = state->isBackwards();
+
+        // The SampleBuffer can play a given sample with increased or decreased pitch. However, only
+        // samples that contain a tone that matches the default base note frequency of 440 Hz will
+        // produce the exact requested pitch in [Hz].
+        const double freqFactor = (double) freq / (double) m_frequency *
+            m_buffer->sampleRate() / Engine::audioEngine()->processingSampleRate();
+
+        // calculate how many frames we have in requested pitch
+        const auto totalFramesForCurrentPitch = static_cast<f_cnt_t>((endFrame - startFrame) / freqFactor);
+
+        if (totalFramesForCurrentPitch == 0)
+        {
+            return false;
+        }
+
+
+        // this holds the index of the first frame to play
+        f_cnt_t playFrame = std::max(state->m_frameIndex, startFrame);
+
+        if (loopMode == LoopMode::LoopOff)
+        {
+            if (playFrame >= endFrame || (endFrame - playFrame) / freqFactor == 0)
```

Comment:
This code is mostly pulled from the current ``SampleBuffer`` class. I plan to lightly refactor this when I have time, but yeah, we probably could just make this a switch statement. Dividing by ``freqFactor`` also seems unnecessary, because for ``(endFrame - playFrame) /  freqFactor == 0`` to return ``true``, ``endFrame`` and ``playFrame`` would've been equal, which was already considered in the first condition, ``playFrame >= endFrame``.

We should also probably have some assertions or return early to avoid dividing by zero errors in ``freqFactor`` (unlikely to happen, but the code is then more robust).